### PR TITLE
Minor code fixes

### PR DESCRIPTION
**Pull request by Fruchtzwerg94**
Used default constructors;
Passed const references;
Removed unused const;
Used empty();
Removed unused virtual;
Used nullptr;

Cheers :-)

### CODE REVIEW

File: src/scn/BlockchainManager/BlockchainManager.h [hunk 1]
```diff
@@ -42,22 +42,22 @@ namespace scn {
...
                 IMiner& miner,
                 bool initial_fetch = true,
                 ISynchronizedTimer& sync_timer = static_sync_timer);
         virtual ~BlockchainManager();
 
-        virtual void baselineBlockReceivedCallback(const peer_id_t& peer_id, std::shared_ptr<const BaselineBlock> block, bool reply);
+        virtual void baselineBlockReceivedCallback(const peer_id_t& peer_id, const std::shared_ptr<const BaselineBlock>& block, bool reply);
 
-        virtual void collectionBlockReceivedCallback(const peer_id_t& peer_id, std::shared_ptr<const CollectionBlock> block, bool reply);
+        virtual void collectionBlockReceivedCallback(const peer_id_t& peer_id, const std::shared_ptr<const CollectionBlock>& block, bool reply);
```
Comment (SwabianCoin) [inline src/scn/BlockchainManager/BlockchainManager.h:57]: I don't really see the sense of passing a const reference to the shared_ptr here (except minimal performance gains).
I personally prefer here to pass the shared_ptr by value to allow the callback to process the blocks asynchronously (the blocks may get pretty big).
See here: https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer

File: src/scn/BlockchainManager/ICycleState.h [hunk 2]
```diff
@@ -33,19 +33,19 @@ namespace scn {
...
 
         virtual bool onCycle() = 0; //return true if there is still something to do
 
         virtual void onExit() = 0;
 
-        virtual void blockReceivedCallback(const peer_id_t& peer_id, std::shared_ptr<const BaselineBlock> block, bool reply) {}
+        virtual void blockReceivedCallback(const peer_id_t& peer_id, const std::shared_ptr<const BaselineBlock>& block, bool reply) {}
 
-        virtual void blockReceivedCallback(const peer_id_t& peer_id, std::shared_ptr<const CollectionBlock> block, bool reply) {}
+        virtual void blockReceivedCallback(const peer_id_t& peer_id, const std::shared_ptr<const CollectionBlock>& block, bool reply) {}
```
Comment (SwabianCoin) [inline src/scn/BlockchainManager/ICycleState.h]: Same as in BlockchainManager.h, I prefer passing the shared_ptr by value here.

**Review by SwabianCoin — CHANGES_REQUESTED** [review]
Thanks for the pull request. There is a lot of good stuff in there.

The only I thing I don't want to merge at the moment is the shared_ptr passing by const reference (see comments in code). Can you remove that in the pull request or explain why you think it's better than the existing implementation?

Greetz

**Comment by Fruchtzwerg94** [comment]
Passing a shared pointer as reference brings up a huge performance increase, see
[Should we pass a shared_ptr by reference or by value?](https://stackoverflow.com/questions/3310737/should-we-pass-a-shared-ptr-by-reference-or-by-value)
Typically there are no reasonst don't passing the pointer not as reference unless you are using it in different threads / working asynchronously. Since this is the case, thats absolutely okay. Thanks for the intend review!

File: src/scn/BlockchainManager/BlockchainManager.h [hunk 3]
```diff
@@ -42,22 +42,22 @@ namespace scn {
...
+                const private_key_t& our_private_key,
                 Blockchain& blockchain,
                 IP2PConnector& p2p_connector,
                 IMiner& miner,
                 bool initial_fetch = true,
                 ISynchronizedTimer& sync_timer = static_sync_timer);
         virtual ~BlockchainManager();
 
-        virtual void baselineBlockReceivedCallback(const peer_id_t& peer_id, std::shared_ptr<const BaselineBlock> block, bool reply);
+        virtual void baselineBlockReceivedCallback(const peer_id_t& peer_id, const std::shared_ptr<const BaselineBlock>& block, bool reply);
```
Comment (SwabianCoin) [inline src/scn/BlockchainManager/BlockchainManager.h:55]: Thank you for the changes and the explanation regarding shared pointers. I am almost ready to merge it now :-)
Just one more thing here: The baselineBlockReceivedCallback in BlockchainManager still changes the shared_ptr to const share_ptr&, probably you just missed it by accident.
Thanks!

**Review by SwabianCoin — CHANGES_REQUESTED** [review]

**Pull request merged**